### PR TITLE
Log Segment Trim

### PR DESCRIPTION
**Pull request by Maithem**

### CODE REVIEW

**Comment by Maithem** [comment]
Two things left: Compact on prefix trims and tests. Working on them now.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **316cf8d**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #639 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/nqZ6S0FIur4vVSUbSCgPyj0WBIoLV4mE)

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java [hunk 1]
```diff
@@ -126,6 +124,9 @@ public LogUnitServer(ServerContext serverContext) {
 
         MetricRegistry metrics = serverContext.getMetrics();
         MetricsUtils.addCacheGauges(metrics, metricsPrefix + "cache.", dataCache);
+
+        Runnable task = () -> streamLog.compact();
+        compactor = scheduler.scheduleAtFixedRate(task, 10,45, TimeUnit.MINUTES);
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: Make compact an api or a periodic task ?

File: test/src/test/java/org/corfudb/infrastructure/log/StreamLogFilesTest.java [hunk 2]
```diff
@@ -366,4 +371,41 @@ public void testGetGlobalTail() {
...
+
+        Set<String> fileNames = new HashSet(Arrays.asList(afterTrimFiles));
+        for (long x = endSegment + 1; x < numSegments; x++) {
+            String logFile = Long.toString(x) + ".log";
+            String trimmedLogFile = StreamLogFiles.getTrimmedFilePath(logFile);
+            String pendingLogFile = StreamLogFiles.getPendingTrimsFilePath(logFile);
+
+            assertThat(fileNames).contains(logFile);
+            assertThat(fileNames).contains(trimmedLogFile);
+            assertThat(fileNames).contains(pendingLogFile);
```
Comment (Maithem) [inline test/src/test/java/org/corfudb/infrastructure/log/StreamLogFilesTest.java]: do reads after this and check for trimmed addresses

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java (hunk 1 above)
Comment (no2chem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: probably both?

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/InMemoryStreamLog.java [hunk 3]
```diff
@@ -63,6 +63,7 @@ public synchronized void append(LogAddress logAddress, LogData entry) {
         globalTail.getAndUpdate(maxTail -> entry.getGlobalAddress() > maxTail ? entry.getGlobalAddress() : maxTail);
     }
 
+    // CHANGE INEQUALITY
```
Comment (no2chem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/InMemoryStreamLog.java]: What do you mean change inequality here?

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java (hunk 1 above)
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: done

**Comment by Maithem** [comment]
@no2chem Done. I think its good to go. Look at the readingTrimmedAddress test, it should be working now.

**Comment by no2chem** [comment]
ok, merging in as soon as it passes.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **2302f4c**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #639 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/xOGEtC14c4awVMYCVnjjLTa6lT1YdeMu)

**Pull request merged**